Mathlib/Data/Int/GCD.lean

```diff
@@ -511,255 +513,269 @@ theorem pow_gcd_eq_one {M : Type _} [Monoid M] (x : M) {m n : ℕ} (hm : x ^ m =
+
+theorem nat_gcd_helper_1 (d x y a b u v tx ty : ℕ) (hu : d * u = x) (hv : d * v = y)
+    (hx : x * a = tx) (hy : y * b = ty) (h : tx + d = ty) : Nat.gcd x y = d :=
+  (Nat.gcd_comm _ _).trans <| nat_gcd_helper_2 _ _ _ _ _ _ _ _ _ hv hu hy hx h
+#align tactic.norm_num.nat_gcd_helper_1 Tactic.NormNum.nat_gcd_helper_1
+
+--Porting note: the `simp only` was not necessary in Lean3.
+theorem nat_lcm_helper (x y d m n : ℕ) (hd : Nat.gcd x y = d) (d0 : 0 < d) (xy : x * y = n)
+    (dm : d * m = n) : Nat.lcm x y = m :=
+  mul_right_injective₀ d0.ne' <| by simp only; rw [dm, ← xy, ← hd, Nat.gcd_mul_lcm]
+#align tactic.norm_num.nat_lcm_helper Tactic.NormNum.nat_lcm_helper
+
+theorem not_coprime_helper {x y d : Nat} (h : Nat.gcd x y = d) (h' : Nat.beq d 1 = false) :
+    ¬ Nat.coprime x y :=
+  by cases h; exact fun h'' => Nat.ne_of_beq_eq_false h' (Nat.coprime_iff_gcd_eq_one.mp h'')
+
+theorem int_gcd_helper {x y : ℤ} {x' y' d : ℕ}
+    (hx : x.natAbs = x') (hy : y.natAbs = y') (h : Nat.gcd x' y' = d) :
+    Int.gcd x y = d := by subst_vars; rw [Int.gcd_def]
+
+theorem int_lcm_helper {x y : ℤ} {x' y' d : ℕ}
+    (hx : x.natAbs = x') (hy : y.natAbs = y') (h : Nat.lcm x' y' = d) :
+    Int.lcm x y = d := by subst_vars; rw [Int.lcm_def]
+
+#noalign tactic.norm_num.nat_coprime_helper_zero_left
+#noalign tactic.norm_num.nat_coprime_helper_zero_right
+#noalign tactic.norm_num.nat_coprime_helper_1
+#noalign tactic.norm_num.nat_coprime_helper_2
+#noalign tactic.norm_num.nat_not_coprime_helper
+#noalign tactic.norm_num.int_gcd_helper''
+#noalign tactic.norm_num.int_gcd_helper_neg_left
+#noalign tactic.norm_num.int_gcd_helper_neg_right
+#noalign tactic.norm_num.int_lcm_helper
+#noalign tactic.norm_num.int_lcm_helper_neg_left
+#noalign tactic.norm_num.int_lcm_helper_neg_right
+
+open Qq Lean Elab.Tactic Mathlib.Meta.NormNum
+
+theorem isNat_gcd : {x y nx ny z : ℕ} →
+    IsNat x nx → IsNat y ny → Nat.gcd nx ny = z → IsNat (Nat.gcd x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+theorem isNat_lcm : {x y nx ny z : ℕ} →
+    IsNat x nx → IsNat y ny → Nat.lcm nx ny = z → IsNat (Nat.lcm x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+theorem isNat_coprime : {x y nx ny : ℕ} →
+    IsNat x nx → IsNat y ny → Nat.coprime nx ny → Nat.coprime x y
+  | _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, h => h
+
+theorem isNat_not_coprime : {x y nx ny : ℕ} →
+    IsNat x nx → IsNat y ny → ¬ Nat.coprime nx ny → ¬ Nat.coprime x y
+  | _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, h => h
+
+theorem isInt_gcd : {x y nx ny : ℤ} → {z : ℕ} →
+    IsInt x nx → IsInt y ny → Int.gcd nx ny = z → IsNat (Int.gcd x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+theorem isInt_lcm : {x y nx ny : ℤ} → {z : ℕ} →
+    IsInt x nx → IsInt y ny → Int.lcm nx ny = z → IsNat (Int.lcm x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+/-- Supporting definition for `proveNatGCD`. Returns the GCD and an equality proof. -/
+def proveNatGCD' (x y : ℕ) : MetaM ((d : ℕ) × Q(Nat.gcd $x $y = $d)) := do
```

Comment:
why `MetaM`?

Reply:
Ha, didn't notice that was unnecessary. Thanks